### PR TITLE
Add dialogs for successful & failed storage migration

### LINKED ISSUE #13131: STOR-05 Simple UI element for when the migration is completed
* This may occur when the user has the app closed 

Accepted Proposal: Toast + Notification

### PR DESCRIPTION
**Pull request by oakkitten**
### Purpose / Description
See #13131. Also see the message of the ~~last~~ commits. Publishing early as requested. ~~Draft because:~~
* this needs to be self-reviewed;
* ~~strings need to be extracted;~~
* ~~`migration_successful` message should be changed to reiterate the bit about the possibility of losing the collection—needs suggestions;~~
* ~~it's midnight and I need to sleep.~~

~~<u>**Ready to be reviewed.**</u> Will fix up the above issues tomorrow if no-one else does.~~ Edit: no longer a draft

### Fixes
Closes #13131. Supersedes and closes #13417.

### Approach
See the messages of the ~~last~~ commits.

### How Has This Been Tested?

Tested on my device.

**NOTE that there are at least three behavior paths to test here: success while app is visible, success while it is not but focused after, success after it is not, killed and reopened, and failure paths for these three cases.**

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [ ] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — APPROVED** [review]

**Comment by oakkitten** [comment]
Updated to:
* extract & change strings;
* dismiss migration progress dialog when done—on of the new dialogs will be shown instead;

This adds a “Learn more and get help” link to the migration failed dialog. This link will need to changed to a proper one once https://github.com/ankidroid/ankidroiddocs/pull/106 is in.

The migration failed dialog message is very generic. It could be customized for certain exceptions, e.g. if we know that the problem is running out of space, we could tell the user that freeing up some space is exactly what they need.

Another major issue (not brought in by this PR) is that after migration failed we are stuck in a kind of a limbo; the migration is considered `Done` (`Failed`, but `Done`) by the service, but `userMigrationIsInProgress` is going to return `true`. I suppose we want to have it return something along “migration paused to to a failure”, and have the UI reflect this state. You can see the problem in the video below: after “Migration failed” dialog is dismissed, the progress button remains rotating (ignore that it doesn't get to actually moving files), and there's no way to resume migration.

Another issue is that as you can see in the video the dialog reads “...after failing with the error: Migration failed”. Apparently this is what we get from `MigrateUserData` instead of the actual exception.

https://user-images.githubusercontent.com/1710718/233791378-77993f48-a762-464e-880b-3f7e808fe565.mp4

**Comment by mikehardy** [comment]
> This adds a “Learn more and get help” link to the migration failed dialog. This link will need to changed to a proper one once https://github.com/ankidroid/ankidroiddocs/pull/106 is in.

it is live now: https://ankidroid.org/storage-migration-error.html

We will always need a generic error message won't we? So having *a* generic message (even if it is the only message at first) seems like a reasonable beginning and not wasted translation effort.

If the migration fails and re-running it does not un-fail it then the user has to go to the storage migration error and probably uninstall / restore from backup? So this state could use some better UI perhaps but not sure how UX could really be improved except perhaps via more specific error messages (the case of running out of space is likely the most important one that users can actually + easily fix...)

File: AnkiDroid/src/main/res/values/constants.xml [hunk 1]
```diff
@@ -165,6 +165,8 @@
     <string name="link_scoped_storage_faq">https://github.com/ankidroid/Anki-Android/wiki/Storage-Migration-FAQ</string>
     <string name="link_install_non_play_store_install">https://github.com/ankidroid/Anki-Android/wiki/Full-Storage-Access</string>
     <string name="link_custom_sync_server_help_learn_more_en">https://docs.ankidroid.org/#_custom_sync_server</string>
+    <!-- TODO BEFORE-RELEASE Insert an actual link here. See https://github.com/ankidroid/ankidroiddocs/pull/106 -->
+    <string name="link_migration_failed_dialog_learn_more_en">https://docs.ankidroid.org/</string>
```
Comment (mikehardy) [inline AnkiDroid/src/main/res/values/constants.xml:169]:
```suggestion
    <string name="link_migration_failed_dialog_learn_more_en">https://docs.ankidroid.org/storage-migration-error.html</string>
```

**Review by mikehardy — APPROVED** [review]
LGTM - proposed actual URL for help doc now that it exists and is live
I will commit that change since this thing is a squash anyway

**Comment by mikehardy** [comment]
Strings are sufficiently clear that strings are *not a blocker*. Ready to go from strings perspective.

File: AnkiDroid/src/main/res/values/constants.xml (hunk 1 above)
Comment (oakkitten) [reply AnkiDroid/src/main/res/values/constants.xml]: Might as remove the TODO there

**Comment by oakkitten** [comment]
> We will always need a generic error message won't we? So having a generic message (even if it is the only message at first) seems like a reasonable beginning and not wasted translation effort.

I'm not sure I follow, is this about the “...after failing with the error: Migration failed”? We should have a non-generic error message there, it's just not exposed to us because of how `AggregateException` works. It probably also will not print an useful traceback. Anyway, this is for another issue.

> If the migration fails and re-running it does not un-fail it then the user has to go to the storage migration error and probably uninstall / restore from backup? So this state could use some better UI perhaps but not sure how UX could really be improved except perhaps via more specific error messages (the case of running out of space is likely the most important one that users can actually + easily fix...)

You can't re-run manually as the UI is just broken in case of failure. But this can be left for another issue as well

**Comment by mikehardy** [comment]
This is a squash merge anyway so I just brutally merged in HEAD to fix a conflict in constants.xml regarding the addition of the link for migration failure so I can get this in

**Pull request merged**